### ISSUE TITLE
Add B3 for ocgrpc

### ISSUE THREAD

**Issue by terinjokes** [census-instrumentation/opencensus-go#666]
In February, support for B3 propagation was added for HTTP clients and servers in #380. It is common in B3 environments to also use these same headers in GRPC metadata, as described in [b3-propagation](https://github.com/openzipkin/b3-propagation/blob/master/README.md#grpc-encoding).

I'm keen to try using OpenCensus in my project, but I need to maintain tracing compatibility with other services over both HTTP and GRPC. At this time, the ocgrpc plugin does now allow for alternative propagation formats.

**Comment by terinjokes** [comment]
> This might be a problem if you are depending on a sidecar or load balancer that only understands gRPC B3 but shouldn't be a problem otherwise.

We do this, as well as communicating over GRPC to services that have no knowledge of OpenCensus, but expect B3 propagation over GRPC.

**Comment by rakyll** [comment]
I see. Let's wait for @bogdandrutu's opinions on this.

**Comment by bogdandrutu** [comment]
@terinjokes I have some questions:
* what languages do you need this?
* Do you expect to always send the B3 propagation format to the next service or do you want to have this as a configuration?

**Comment by terinjokes** [comment]
I'm primarily writing Go. Many of the services I interact with (which may be using OpenZipkin instead of OpenCensus) are also written in Go, though some are not.

To prevent fragmentation of distributed tracing data, we've specified using B3 propagation between all services, where this is a reasonable ask. As linked above, this is defined for HTTP (headers) and GRPC (metadata), and we use similarly named fields when enqueing in a message queue.

B3 was chosen for it's broad support between applications, languages, and frameworks.

**Comment by rakyll** [comment]
Talked to @bogdandrutu offline and we think providing an option for B3 propagation is fair and required to support legacy systems. We will keep using the binary format and add B3 headers if users wanted them.

**Comment by adriancole** [comment]
I would set expectations to make the grpc binary format possible vs
expecting a switch. Switch is very hard in brown field especially as
libraries that do B3 with grpc are not in openzipkin's org and we dont have
strict control of people's site deployments. It is more complicated even to
attempt to switch.

I think carrot will be that whatever binary format.. that this buys more
than just serialization overhead as the cost to push people towards
anything they arent currently focused on is quite high.

For example if the binary format is somehow related to trace context or
something such that would be a decent sell as the folks who are less likely
to move may be on account of not being quite concerned about the cost to
marshal hex which to many is negligible as it is far under a microsecond
and http header compression should take care of the repeated B3 keys.

Basically I would tie this leverage towards binary format into something
not singularly focused on grpc rather a bigger win that helps with other
frameworks too.

My 2p.

**Comment by adriancole** [comment]
By the way, in Brave (java library), I am happy to help investigate a
framework-specific (grpc in this case) override towards the current version
of grpc format. That way using the format internal to grpc doesnt bleed
into other systems like http.

Current practice in just about all libraries is to be consistent with one
format vs having different ones per platform. It is more complicated but
still willing to help. At the very least we could read incoming grpc
specific trace format, maybe stashing that info for outgoing or dual
propagating or something.

**Comment by adriancole** [comment]
>
> That way using the format internal to grpc doesnt bleed
> into other systems like http.
>
> The gRPC format is not leaking into other systems because we immediately
> convert the header to a SpanContext and propagate it in a context.Context
> object inside the project. Are you talking about something else?
>
Saying if this is special-cased for grpc, we should be as clear as
possible. Early experiments seem "fine" for interop with the binary format,
just it is a lot easier to contain if pinned to grpc vs all transports,
especially with in-flight work on trace context possibly affecting us later.

Here's an example experiment where I only read/write grpc binary format
when .. well in grpc
https://github.com/openzipkin/brave/pull/688

**Comment by terinjokes** [comment]
@rakyll I think sending tracing metadata in both formats is a reasonable alternative.

I'll likely need to coordinate with the middleware (in this case, Envoy) to ensure when it creates child spans, it also updates the native format.

**Comment by ramonza** [comment]
I need this for Istio trace support. I will probably start working on it in the next day or so.

**Comment by ramonza** [comment]
For now, it uses B3 as the propagation format. I might try to support other propagation formats as a follow up but it requires changes in a number of different Istio components so want to just go with B3 for now.

**Comment by rakyll** [comment]
Duplicate of https://github.com/census-instrumentation/opencensus-specs/pull/136.

**Comment by ramonza** [comment]
We can use this to track implementation once the spec issue is accepted.

**Comment by adriancole** [comment]
so B3 gets the number 666 which while lucky in chinese is somewhat less so
in other cultures ;)

**Comment by fho** [comment]
We also require support for B3 propagation especially on the GRPC server-side.
We have GRPC-Clients in different languages (e.g. python, php, nodejs) which propagate the tracing information in B3 format to the GRPC-Servers.
We can not change all applications at once to use the binary format for GRPC propagation.
Support for B3 format would allow us to migrate to opencensus in the Go-GRPC-Servers and then migrate step by step the other apps to use opencensus.

**Comment by jcxplorer** [comment]
I've made a small [proof of concept](https://github.com/jcxplorer/opencensus-go/commit/028a4c806b7d8e24a8316bcd85e27651b90a1d65) for this and tested it with Istio. It's a small change, so if there's interest in merging it, I'd be happy to clean it up, add tests, and make it optional in `ocgrpc.ClientHandler`. I don't think we need support in `ServerHandler`, because we still pass the current `grpc-trace-bin`. At least for distributed tracing with Istio, this is enough.

Any of the current maintainers willing to accept a PR?

**Comment by tmc** [comment]
Is there an opentelemetry equivalent for this?
